### ISSUE TITLE
Calling addSelect('field') with no earlier select('*') or addSelect('*') results in almost empty object (with only 'field' retrieved)

### ISSUE THREAD

**Issue by artur-ciesielski** [yiisoft/yii2#12249]
### Problem:
When using addSelect() on a query with no previous select('*') and addSelect('*') methods called, the result will be an empty object with only the fields from addSelect(). It's probably the result of this:
```
if ($this->select === null) {
    $this->select = $columns;
} else {
    $this->select = array_merge($this->select, $columns);
} // where columns is the argument passed to addSelect()
```
I am willing to cobble together a pull request for this later this week, if the idea is acceptable (I assume there is a reason for that behavior).

### What steps will reproduce the problem?
Using any query with addSelect() without using select('*') or addSelect('*') beforehand.

### What is the expected result?
addSelect() should only expand the list of fields selected, not assume no fields should be selected by default. After performing the query an object with selected all fields AND the fields specified in addSelect().

### What do you get instead?
Object with all fields NULL except for those specified in addSelect().

### Additional info

| Q                | A
| ---------------- | ---
| Yii version | 2.0.9
| PHP version | 7.0.9
| Operating system | Ubuntu 16.04 x86_64

**Comment by samdark** [comment]
I think it's expected to work like that. @yiisoft/core-developers do you remember if it's true or not?

**Comment by artur-ciesielski** [comment]
@samdark Yeah, I assume so. The question is is there a basis for this reasoning or should we make this an RFC, perhaps?

**Comment by Alex-Code** [comment]
I think the current behavior is correct.

Here I can add `col2` or `col3` to the select without needing to know if the select has already been specified. If `$condition1` is false I don't want to return all columns with `col2` or `col3`.
e.g.
```php
if ($condition1) {
    $query->select('col1');
}

if ($condition2) {
    $query->addSelect('col2');
} else if ($condition3) {
    $query->addSelect('col3');
}
```

**Comment by artur-ciesielski** [comment]
@Alex-Code I see your point.

Right now it's consistent with `andWhere()` and the like, as in it assumes `NULL` when the condition is not set beforehand.

On the other hand, by default `ActiveRecord::find()` returns all columns when `select()` is not specified. This makes using `addSelect()` in it's current state kind of counter-intuitive and requires using `addSelect(['*', 'extraColumn'])` every time I want to simply add another column to `SELECT` part of the query.

Also, this breaks when using a custom Query class, as every time I would like to use a query method with `addSelect()` I have to check for conflicting `addSelect()` calls in other methods in the custom Qeury class and work around them.

**Comment by samdark** [comment]
You need to call `->select('*')` prior to `addSelect()` if you need behavior you want.

**Comment by artur-ciesielski** [comment]
@samdark Yeah, that's what I am doing. It's sad that I have to do that every single time though, since I don't want to put `select('*')` in my Query class methods.

I think we can close this, unless anyone has any more insight on the matter to share.

**Comment by samdark** [comment]
OK.

**Comment by Alex-Code** [comment]
You could do this in your model
```php
/**
* @return ActiveQuery
*/
public static function find()
{
    $query = new ActiveQuery(get_called_class());
    return $query->select('*');
}
```

**Comment by artur-ciesielski** [comment]
@Alex-Code Yeah, that's what I figured out as well. Thanks for your time though.

**Comment by undestroyer** [comment]
I faced this issue right now. I was surprized that:
1. Due to docs,  `select` property means "all by default" [yii\db\Query:58](https://github.com/yiisoft/yii2/blob/05f197825b1e3da9d931b5e620643811e6581bc9/framework/db/Query.php#L58) 
```
 /**
  * @var array the columns being selected. For example, `['id', 'name']`.
  * This is used to construct the SELECT clause in a SQL statement. If not set, it means selecting all columns.
  * @see select()
  */
public $select; 
```
2. Add statement ignores. [yii\db\Query:592](https://github.com/yiisoft/yii2/blob/05f197825b1e3da9d931b5e620643811e6581bc9/framework/db/Query.php#L592)
```
if ($this->select === null) {
    $this->select = $columns;
} else {
    $this->select = array_merge($this->select, $columns);
}
```

Some use cases:

| What I want 	| How do I do 	| Is goal Reached? 	|
|-------------------------------------	|--------------------------------------------------------------------------------	|------------------	|
| Select all default columns 	| ``ArClass::find()->all() // no call select()`` 	| yes 	|
| Select only 2 passed columns 	| ``ArClass::find()->select(['col1', 'col2'])->all()`` 	| yes 	|
| Select 2 passed columns +1 more 	| ``ArClass::find()->select(['col1', 'col2'])->addSelect('col3')->all()`` 	| yes 	|
| Select all default columns + 1 more 	| ``ArClass::find()->addSelect('col3')->all()`` 	| no 	|
| Select all default columns + 1 more 	| ``ArClass::find()->select('*')->addSelect('col3')->all()`` 	| sometimes 	|
| Select all default columns + 1 more 	| ``ArClass::find()->select(ArClass::tableName() . '.*')->addSelect('col3')->all()`` 	| yes 	|


In my opinion, query must select all default columns until I define columns set manually. Add meand adding element to some set. `$select` defined as **all by default**, but when you call `addSelect()` without calling `select()`, that means **default not exists, add this**.

All suggested solutions are workarounds, neither of them solves this semantic problem. Once again, I speek only about semantic problem with word "add" in method name.

**Comment by samdark** [comment]
@undestroyer would you please post it as separate issue? We can possibly fix it in 2.1.

**Comment by undestroyer** [comment]
@samdark Yes, of course! Thank you for reply.
